File: python/persistenttrafficlight/trafficlight/trafficlight.py
# emitted from framec_v0.10.0
# get include files at https://github.com/frame-lang/frame-ancillary-files
import copy
from framelang.framelang import FrameEvent

class TrafficLight:
    
    def __init__(self, manager):
        
        # Create and intialize start state compartment.
        self.__state = self.__trafficlight_state_Begin
        self._manager = manager
        self.__compartment: 'TrafficLightCompartment' = TrafficLightCompartment(self.__state)
        self.__next_compartment: 'TrafficLightCompartment' = None
        
        # Initialize domain
        
        self.flashColor : str = ""
        
        # Send system start event
        frame_event = FrameEvent(">", None)
        self.__mux(frame_event)
    
    # ===================== Interface Block =================== #
    
    def stop(self,):
        e = FrameEvent("stop",None)
        self.__mux(e)
    
    def tick(self,):
        e = FrameEvent("tick",None)
        self.__mux(e)
    
    def systemError(self,):
        e = FrameEvent("systemError",None)
        self.__mux(e)
    
    def systemRestart(self,):
        e = FrameEvent("systemRestart",None)
        self.__mux(e)
    
    
    
    @staticmethod
    def loadTrafficLight(manager, data):
        
        data._manager = manager
        
        return data
        
    
    
    def marshal(self):
        
        data = copy.deepcopy(self)
        return data
        
    
    # ====================== Multiplexer ==================== #
    
    def __mux(self, e):
        if self.__compartment.state == self.__trafficlight_state_Begin:
            self.__trafficlight_state_Begin(e)
        elif self.__compartment.state == self.__trafficlight_state_Red:
            self.__trafficlight_state_Red(e)
        elif self.__compartment.state == self.__trafficlight_state_Green:
            self.__trafficlight_state_Green(e)
        elif self.__compartment.state == self.__trafficlight_state_Yellow:
            self.__trafficlight_state_Yellow(e)
        elif self.__compartment.state == self.__trafficlight_state_FlashingRed:
            self.__trafficlight_state_FlashingRed(e)
        elif self.__compartment.state == self.__trafficlight_state_End:
            self.__trafficlight_state_End(e)
        elif self.__compartment.state == self.__trafficlight_state_Working:
            self.__trafficlight_state_Working(e)
        
        if self.__next_compartment != None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            if(next_compartment.forward_event is not None and 
               next_compartment.forward_event._message == ">"):
                self.__mux(FrameEvent( "<", self.__compartment.exit_args))
                self.__compartment = next_compartment
                self.__mux(next_compartment.forward_event)
            else:
                self.__do_transition(next_compartment)
                if next_compartment.forward_event is not None:
                    self.__mux(next_compartment.forward_event)
            next_compartment.forward_event = None
    
    # ===================== Machine Block =================== #
    
    def __trafficlight_state_Begin(self, e):
        if e._message == ">":
            self.startWorkingTimer_do()
            compartment = TrafficLightCompartment(self.__trafficlight_state_Red)
            self.__transition(compartment)
            return
        
    def __trafficlight_state_Red(self, e):
        if e._message == ">":
            self.enterRed_do()
            return
        
        elif e._message == "tick":
            compartment = TrafficLightCompartment(self.__trafficlight_state_Green)
            self.__transition(compartment)
            return
        
        self.__trafficlight_state_Working(e)
        
    def __trafficlight_state_Green(self, e):
        if e._message == ">":
            self.enterGreen_do()
            return
        
        elif e._message == "tick":
            compartment = TrafficLightCompartment(self.__trafficlight_state_Yellow)
            self.__transition(compartment)
            return
        
        self.__trafficlight_state_Working(e)
        
    def __trafficlight_state_Yellow(self, e):
        if e._message == ">":
            self.enterYellow_do()
            return
        
        elif e._message == "tick":
            compartment = TrafficLightCompartment(self.__trafficlight_state_Red)
            self.__transition(compartment)
            return
        
        self.__trafficlight_state_Working(e)
        
    def __trafficlight_state_FlashingRed(self, e):
        if e._message == ">":
            self.enterFlashingRed_do()
            self.stopWorkingTimer_do()
            self.startFlashingTimer_do()
            return
        
        elif e._message == "<":
            self.exitFlashingRed_do()
            self.stopFlashingTimer_do()
            self.startWorkingTimer_do()
            return
        
        elif e._message == "tick":
            self.changeFlashingAnimation_do()
            return
        
        elif e._message == "systemRestart":
            compartment = TrafficLightCompartment(self.__trafficlight_state_Red)
            self.__transition(compartment)
            return
        
        elif e._message == "stop":
            compartment = TrafficLightCompartment(self.__trafficlight_state_End)
            self.__transition(compartment)
            return
        
    def __trafficlight_state_End(self, e):
        if e._message == ">":
            self.stopWorkingTimer_do()
            return
        
    def __trafficlight_state_Working(self, e):
        if e._message == "stop":
            compartment = TrafficLightCompartment(self.__trafficlight_state_End)
            self.__transition(compartment)
            return
        
        elif e._message == "systemError":
            compartment = TrafficLightCompartment(self.__trafficlight_state_FlashingRed)
            self.__transition(compartment)
            return
        
    
    # ===================== Actions Block =================== #
    
    
    # Unimplemented Actions
    
    def enterRed_do(self):
        raise NotImplementedError
    
    def enterGreen_do(self):
        raise NotImplementedError
    
    def enterYellow_do(self):
        raise NotImplementedError
    
    def enterFlashingRed_do(self):
        raise NotImplementedError
    
    def exitFlashingRed_do(self):
        raise NotImplementedError
    
    def startWorkingTimer_do(self):
        raise NotImplementedError
    
    def stopWorkingTimer_do(self):
        raise NotImplementedError
    
    def startFlashingTimer_do(self):
        raise NotImplementedError
    
    def stopFlashingTimer_do(self):
        raise NotImplementedError
    
    def changeColor_do(self,color: str):
        raise NotImplementedError
    
    def startFlashing_do(self):
        raise NotImplementedError
    
    def stopFlashing_do(self):
        raise NotImplementedError
    
    def changeFlashingAnimation_do(self):
        raise NotImplementedError
    
    def log_do(self,msg: str):
        raise NotImplementedError
    
    
    # =============== Machinery and Mechanisms ============== #
    
    def __transition(self, compartment: 'TrafficLightCompartment'):
        self.__next_compartment = compartment
    
    def  __do_transition(self, next_compartment: 'TrafficLightCompartment'):
        self.__mux(FrameEvent("<", self.__compartment.exit_args))
        self.__compartment = next_compartment
        self.__mux(FrameEvent(">", self.__compartment.enter_args))
    

# ===================== Compartment =================== #

class TrafficLightCompartment:

    def __init__(self, state):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = FrameEvent(None, None)
    


# ********************

class TrafficLightController(TrafficLight):
    def __init__(self, manager):
        super().__init__(manager)

    def enterRed_do(self):
        self._manager.enterRed()
    def enterGreen_do(self):
        self._manager.enterGreen()

    def enterYellow_do(self):
        self._manager.enterYellow()

    def enterFlashingRed_do(self):
        pass

    def exitFlashingRed_do(self):
        pass

    def startWorkingTimer_do(self):
        pass

    def stopWorkingTimer_do(self):
        pass

    def startFlashingTimer_do(self):
        pass

    def stopFlashingTimer_do(self):
        pass

    def changeColor_do(self,color):
        pass

    def startFlashing_do(self):
        pass

    def stopFlashing_do(self):
        pass

    def changeFlashingAnimation_do(self):
        pass

    def log_do(self,msg):
        pass

# ********************

File: javaframedemos/PersistentTrafficLight/javaframedemos/PersistentTrafficLight/TrafficLight.java
package javaframedemos.PersistentTrafficLight;

import java.util.*;

class TrafficLight{

    // creating private properties
    private int __state;
    private TrafficLightCompartment _compartment_;
    private TrafficLightCompartment _nextCompartment_;
    protected TrafficLightMom _manager;
    FrameEvent e = new FrameEvent();
    TrafficLight tLight;
    public TrafficLight(TrafficLightMom manager) {
        this.__state = this._sBegin_(e);
        this._manager = manager;
        this._compartment_ = new TrafficLightCompartment(this.__state);
        this._nextCompartment_ = null;
        
        // Send system start event
        e = new FrameEvent(">", null);
        try{
           this._mux_(e);
        }catch(Exception ex){
            //ex.printStackTrace();
        }   
    }

    // state enum
    private enum TrafficLightState {
        BEGIN(0),
        RED(1),
        GREEN(2),
        YELLOW(3),
        FLASHINGRED(4),
        END(5),
        WORKING(6);

        private final int value;

        private TrafficLightState(int value) {
            this.value = value;
        }

        public int getValue() {
            return value;
        }

    }

    // ===================Interface Block===================//

    public void stop()  throws Exception{
        e = new FrameEvent("stop", null);
        this._mux_(e);
    }

    public void tick()  throws Exception{
        e = new FrameEvent("tick", null);
        this._mux_(e);
    }

    public void systemError()  throws Exception{
        e = new FrameEvent("systemError", null);
        this._mux_(e);
    }

    public void systemRestart()  throws Exception{
        e = new FrameEvent("systemRestart", null);
        this._mux_(e);
    }

    public static TrafficLight loadTrafficLight(TrafficLightMomController manager, TrafficLightController data) {
            data._manager = manager;
        return data;
    }

    public TrafficLight marshal() {
        TrafficLight data = this;
        return data;
    }

    // ===================Multiplexer===================//
    public void _mux_(FrameEvent e)  throws Exception{
        if (this._compartment_.state == TrafficLightState.BEGIN.getValue()) {
            this._sBegin_(e);
        } else if (this._compartment_.state == TrafficLightState.RED.getValue()) {
            this._sRed_(e);
        } else if (this._compartment_.state == TrafficLightState.GREEN.getValue()) {
            this._sGreen_(e);
        } else if (this._compartment_.state == TrafficLightState.YELLOW.getValue()) {
            this._sYellow_(e);
        } else if (this._compartment_.state == TrafficLightState.FLASHINGRED.getValue()) {
            this._sFlashingRed_(e);
        } else if (this._compartment_.state == TrafficLightState.END.getValue()) {
            this._sEnd_(e);
        } else if (this._compartment_.state == TrafficLightState.WORKING.getValue()) {
            this._sWorking_(e);
        }

        if (this._nextCompartment_ != null) {
            TrafficLightCompartment nextCompartment = this._nextCompartment_;
            this._nextCompartment_ = null;
            if (nextCompartment._forwardEvent != null &&
                    nextCompartment._forwardEvent._message == ">") {
                this._mux_(new FrameEvent("<", this._compartment_.ExitArgs));
                this._compartment_ = nextCompartment;
                this._mux_(nextCompartment._forwardEvent);
            } else {
                this._doTransition_(nextCompartment);
                if (nextCompartment._forwardEvent != null) {
                    this._mux_(nextCompartment._forwardEvent);
                }
            }
            nextCompartment._forwardEvent = null;
        }
    }

    // =================== Machine Block ===================//

    private int _sBegin_(FrameEvent e) {
        if (e._message == ">") {
            this.startWorkingTimer_do();
            TrafficLightCompartment compartment = new TrafficLightCompartment(TrafficLightState.RED.getValue());
            return this._transition_(compartment);
        }
        return 0;
    }

    private void _sRed_(FrameEvent e)  throws Exception{
        if (e._message == ">") {
            this.enterRed_do();
            return;
        } else if (e._message == "tick") {
            TrafficLightCompartment compartment = new TrafficLightCompartment(TrafficLightState.GREEN.getValue());
            this._transition_(compartment);
            return;
        }
        this._sWorking_(e);

    }

    private void _sGreen_(FrameEvent e)  throws Exception{
        if (e._message == ">") {
            this.enterGreen_do();
            return;
        } else if (e._message == "tick") {
            TrafficLightCompartment compartment = new TrafficLightCompartment(TrafficLightState.YELLOW.getValue());
            this._transition_(compartment);
            return;
        }
        this._sWorking_(e);

    }

    private void _sYellow_(FrameEvent e)  throws Exception{
        if (e._message == ">") {
            this.enterYellow_do();
            return;
        } else if (e._message == "tick") {
            TrafficLightCompartment compartment = new TrafficLightCompartment(TrafficLightState.RED.getValue());
            this._transition_(compartment);
            return;
        }
        this._sWorking_(e);

    }

    private void _sFlashingRed_(FrameEvent e) {
        if (e._message == ">") {
            this.enterFlashingRed_do();
            this.stopWorkingTimer_do();
            this.startFlashingTimer_do();
            return;
        } else if (e._message == "<") {
            this.exitFlashingRed_do();
            this.stopFlashingTimer_do();
            this.startWorkingTimer_do();
            return;
        } else if (e._message == "tick") {
            this.changeFlashingAnimation_do();
            return;
        } else if (e._message == "systemRestart") {
            TrafficLightCompartment compartment = new TrafficLightCompartment(TrafficLightState.RED.getValue());
            this._transition_(compartment);
            return;
        } else if (e._message == "stop") {
            TrafficLightCompartment compartment = new TrafficLightCompartment(TrafficLightState.END.getValue());
            this._transition_(compartment);
            return;
        }

    }

    private void _sEnd_(FrameEvent e) {
        if (e._message == ">") {
            this.stopWorkingTimer_do();
            return;
        }
    }

    private void _sWorking_(FrameEvent e)  throws Exception{
        if (e._message == "stop") {
            TrafficLightCompartment compartment = new TrafficLightCompartment(TrafficLightState.END.getValue());
            this._doTransition_(compartment);
            return;
        } else if (e._message == "systemError") {
            TrafficLightCompartment compartment = new TrafficLightCompartment(TrafficLightState.FLASHINGRED.getValue());
            this._doTransition_(compartment);
            return;
        }
    }

    // =================== Action Block ==================//
    // Unimplemented Actions

    protected void enterRed_do() throws Exception { throw new UnsupportedOperationException(); }
    protected void enterGreen_do() throws Exception { throw new UnsupportedOperationException(); }
    protected void enterYellow_do() throws Exception { throw new UnsupportedOperationException(); }
    protected void enterFlashingRed_do() { throw new UnsupportedOperationException(); }
    protected void exitFlashingRed_do() { throw new UnsupportedOperationException(); }
    protected void startWorkingTimer_do() { throw new UnsupportedOperationException(); }
    protected void stopWorkingTimer_do() { throw new UnsupportedOperationException(); }
    protected void startFlashingTimer_do() { throw new UnsupportedOperationException(); }
    protected void stopFlashingTimer_do() { throw new UnsupportedOperationException(); }
    protected void changeColor_do(String color) { throw new UnsupportedOperationException(); }
    protected void startFlashing_do() { throw new UnsupportedOperationException(); }
    protected void stopFlashing_do() { throw new UnsupportedOperationException(); }
    protected void changeFlashingAnimation_do() { throw new UnsupportedOperationException(); }
    protected void log_do(String msg) { throw new UnsupportedOperationException(); }

    // =============== Machinery and Mechanisms ==============//

    private int _transition_(TrafficLightCompartment compartment) {
        this._nextCompartment_ = compartment;
        return this._nextCompartment_.state;
    }

    private void _doTransition_(TrafficLightCompartment nextCompartment)  throws Exception{
        this._mux_(new FrameEvent("<", this._compartment_.ExitArgs));
        this._compartment_ = nextCompartment;
        this._mux_(new FrameEvent(">", this._compartment_.EnterArgs));
    }

}

// =============== Compartment ==============//
class TrafficLightCompartment {
    int state;

    TrafficLightCompartment(int state) {
        this.state = state;
    }

    HashMap<String, Object> StateArgs = new HashMap<String, Object>();
    HashMap<String, Object> StateVars = new HashMap<String, Object>();
    HashMap<String, Object> EnterArgs = new HashMap<String, Object>();
    HashMap<String, Object> ExitArgs = new HashMap<String, Object>();
    FrameEvent _forwardEvent = new FrameEvent();
}

class TrafficLightController extends TrafficLight {
    TrafficLightController(TrafficLightMom manager) {
        super(manager);
    }
    @Override
    protected void enterRed_do() {
        //System.out.println("RED");
        //this._manager.enterRed_do();
        try {
            this._manager.enterRed();
        } catch (Exception e) {
           // e.printStackTrace();
        }
    }

    @Override
    protected void enterGreen_do() {
        //System.out.println("GREEN");
        try {
            this._manager.enterGreen();
        } catch (Exception e) {
           // e.printStackTrace();
        }
    }

    @Override
    protected void enterYellow_do() {
        //System.out.println("YELLOW");
        try {
            this._manager.enterYellow();
        } catch (Exception e) {
           // e.printStackTrace();
        }
    }

    protected void enterFlashingRed_do() {
    }

    protected void exitFlashingRed_do() {
    }

    protected void startWorkingTimer_do() {
    }

    protected void stopWorkingTimer_do() {
    }

    protected void startFlashingTimer_do() {
    }

    protected void stopFlashingTimer_do() {
    }

    protected void changeColor_do(String color) {
    }

    protected void startFlashing_do() {
    }

    protected void stopFlashing_do() {
    }

    protected void changeFlashingAnimation_do() {
    }

    protected void log_do(String msg) {
    }

}